apps/civicsignalblog/migrations/20241011_104214_migration.ts

```diff
@@ -0,0 +1,50 @@
+import { MigrateUpArgs, MigrateDownArgs } from "@payloadcms/db-mongodb";
+import mongoose from "mongoose";
+
+const mongoURL = process.env.MONGO_URL;
+const oldCollectionName = "pages";
+const newCollectionName = "research-pages";
+
+export async function up({ payload }: MigrateUpArgs): Promise<void> {
+  // Connect to the database
+  await mongoose.connect(mongoURL);
+  const db = mongoose.connection.db;
+
+  // We will need the admin database for rename operation see https://www.mongodb.com/docs/manual/reference/command/renameCollection/
+  const adminDb = mongoose.connection.getClient().db("admin");
+
+  const oldCollectionVersionName = `_${oldCollectionName}_versions`;
+  const newCollectionVersionName = `_${newCollectionName}_versions`;
+
+  // Drop the new collection, since it has nothing and we cant rename the old collection to new name when this exists
+  await db.command({ drop: newCollectionName });
+  payload.logger.info(`Dropped the existing ${newCollectionName} collection.`);
```

Comment:
Can we do this **without** dropping _any_ db? May be warn and exit if new collection already exists?

Reply:
mmh there's no clean solution that I can quickly think of...we'll always have to drop a collection from the db since we are trying to avoid name collision when we do the rename operation.

If we warn and exit it would not solve our problem (currently we already have `research-pages` which have no data, and we also have `pages` which have data and cant be viewed in payload cms/front-end, thus we don't really need `research-pages` and we can safely drop them.

Alternatively, we could rename the collections to something else like `_{$newCollectionName}` if we really want to avoid dropping collections

Reply:
Yes it won't **solve** the problem but it will tell us that _something_ went wrong. Basically, when it comes to managing data, you don't want automatic data deletion anywhere in the process, **especially** in `PROD`. Someone must 👁️ and give the 👍🏽 to delete (preferably after a careful and thorough review).

... but to make sure we're all on the same page, First time migrations are run:
1. the script will **find** old collection **but not** new collection (no need to delete).
2. If successful, at the end of migration, old collection would have been renamed to new collection.

If someone runs the migration again:
1. the script **will not find** old collection **but** there will be new collection.
2. the script will delete this new collection.
3. since there is no old collection, there is nothing to rename.


I'll sleep better an night if we issue warning and exit in ☝🏽 situation without deleting anything.

Reply:
It is slightly different:

For the first time the migration is run:
1. The script will find both the old and new collection, the reason being that Payload has already automatically created a new collection after it found the new slug i.e. `research-pages` but didn't delete the old collection which has `pages`.

1. So running the rename the operation without deleting the new-collection will always fail due to name collision.

Additionally, Payload keeps a [record](https://payloadcms.com/docs/database/migrations#status) of executed migrations and will not execute them when someone re-runs the migrations unless they first intentionally rollback the execution status

![Screenshot 2024-10-15 at 14 59 37](https://github.com/user-attachments/assets/b139865b-4301-4676-914b-e8d52b8324a4)

I believe the best approach that will improve the migration is to prompt the user for confirmation that the two collections will be dropped before proceeding with the execution, (we can also include the number of records found in each of the collections that will be dropped, so that someone can make an informed decision while knowing the potential consequences of running the migration).

Reply:
I hear you but still no... I'm not comfortable with deleting data in a script like this.

If the new collection is already created, the best approach would be to copy items from old collection into new collection using item `slug` as unique key for detecting duplicates (if Mongo supports that).

This means all documents in old collection that are not in the new collection will be copied over. If there is a `slug` clash, the script should log a warning but not override what's in the new collection already.

Someone can review and delete the old collection when they're happy that everything has been migrated into the new collection.

Reply:
Cool, let me go with this approach !

Reply:
@kilemensi, implementation using the new approach is done, luckily Payload does mark the `slug` as a unique field thus we can easily detect duplicates and log an error. I believe this should be in good order !